x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -49,56 +58,184 @@
+        this(organicQuery, null, Arrays.asList(documents));
+    }
+
+    /**
+     * Creates a new PinnedQueryBuilder
+     */
+    public PinnedQueryBuilder(QueryBuilder organicQuery, List<String> ids, List<Item> documents) {
+        if (organicQuery == null) {
+            throw new IllegalArgumentException("[" + NAME + "] organicQuery cannot be null");
+        }
+        this.organicQuery = organicQuery;
+        if (ids == null && documents == null) {
+            throw new IllegalArgumentException("[" + NAME + "] ids and documents cannot both be null");
+        }
+        if (ids != null && documents != null) {
+            throw new IllegalArgumentException("[" + NAME + "] ids and documents cannot both be used");
+        }
+        if (ids != null) {
+            if (ids.size() > MAX_NUM_PINNED_HITS) {
+                throw new IllegalArgumentException(
+                    "[" + NAME + "] Max of " + MAX_NUM_PINNED_HITS + " ids exceeded: " + ids.size() + " provided."
+                );
+            }
+            LinkedHashSet<String> deduped = new LinkedHashSet<>();
+            for (String id : ids) {
+                if (id == null) {
+                    throw new IllegalArgumentException("[" + NAME + "] id cannot be null");
+                }
+                if (deduped.add(id) == false) {
+                    throw new IllegalArgumentException("[" + NAME + "] duplicate id found in list: " + id);
+                }
+            }
+        }
+        if (documents != null) {
+            if (documents.size() > MAX_NUM_PINNED_HITS) {
+                throw new IllegalArgumentException(
+                    "[" + NAME + "] Max of " + MAX_NUM_PINNED_HITS + " documents exceeded: " + documents.size() + " provided."
+                );
+            }
+            LinkedHashSet<Item> deduped = new LinkedHashSet<>();
+            for (Item document : documents) {
+                if (document == null) {
+                    throw new IllegalArgumentException("[" + NAME + "] document cannot be null");
+                }
+                if (deduped.add(document) == false) {
+                    throw new IllegalArgumentException("[" + NAME + "] duplicate document found in list: " + document);
+                }
+            }
+        }
+        this.ids = ids;
+        this.documents = documents;
     }
 
     /**
      * Read from a stream.
      */
     public PinnedQueryBuilder(StreamInput in) throws IOException {
         super(in);
-        ids = in.readStringList();
+        ids = in.readOptionalStringList();
+        documents = in.readBoolean() ? in.readList(Item::new) : null;
         organicQuery = in.readNamedWriteable(QueryBuilder.class);
     }
 
     @Override
     protected void doWriteTo(StreamOutput out) throws IOException {
-        out.writeStringCollection(this.ids);
+        out.writeOptionalStringCollection(this.ids);
+        if (documents == null) {
```

Comment:
We'll need to add a check to make sure we don't send 'documents' to nodes on older versions, since they aren't expecting this key. We can use `out.getVersion().onOrAfter(...)` to check the version. The input side using `StreamInput` will need a similar check, since old nodes will send data in the old format.

Reply:
Ah, that makes sense. How is the version number handled when backporting? I used `getVersion().before(Version.V_8_0_0)` to gate the changes, but I assume `Version.V_8_0_0` changes if we backport?

Reply:
Right, when backporting we'll switch it to something like `getVersion().before(Version.V_7_15_0)`, then once that's merged we'll immediately change this on master to be `Version.V_7_15_0` too. I'm also happy to handle the backport since it can be tricky.